File: DongHeonKim/5주차/BJ_15810_풍선_공장.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();
        int M = sc.nextInt();

        int[] time = new int[N];
        for (int i = 0; i < N; i++) {
            time[i] = sc.nextInt();
        }

        long left = 1;
        long right = 1_000_000L * M;

        long answer = right;

        while (left <= right) {
            long mid = (left + right) / 2;
            long total = 0;

            for (int i = 0; i < N; i++) {
                total += mid / time[i];
                if (total >= M) break;
            }

            if (total >= M) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        System.out.println(answer);
    }
}

File: Ajeong-Im/5주차/BJ_15810_풍선_공장.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const [N, M] = input[0].split(" ").map(Number);
const A = input[1].split(" ").map(Number);

let lo = 1;
let hi = Math.max(...A) * M;
let answer = hi;

while (lo <= hi) {
  const mid = Math.floor((lo + hi) / 2);

  // mid 시간 동안 만들 수 있는 풍선 총 개수
  const total = A.reduce((sum, time) => sum + Math.floor(mid / time), 0);

  if (total >= M) {
    answer = mid;
    hi = mid - 1; // 더 빠른 시간 가능한지 확인
  } else {
    lo = mid + 1; // 더 많은 시간이 필요함
  }
}

console.log(answer);
